tomcat-job-type/base.Dockerfile

```diff
@@ -1,18 +1,23 @@
 FROM gradle:8.4.0-jdk21 as build
 WORKDIR /src
 COPY ./app ./app
+COPY ./prometheus ./prometheus
 COPY settings.gradle.kts ./
 RUN gradle war
 
+
 #---------------------------
 FROM tomcat:10.1.15-jre21-temurin
 
+COPY server.xml "${CATALINA_HOME}/conf/server.xml"
+
 # This war contains Health component, which will serve /health, /live and /ready, on root endpoint.
 COPY --from=build /src/app/build/libs/app.war "${CATALINA_HOME}/webapps/ROOT.war"
 
+COPY --from=build /src/prometheus/build/libs/prometheus.war "${CATALINA_HOME}/webapps/prometheus.war"
```

Comment:
If prometheus is a separate application (that's how I perceive war files, but maybe I'm wrong), is it possible to serve metrics that are being set in a Job class (from app.war)?

Reply:
Yes and no. Yes, because Job class is running under the same endpoint as Prometheus, like `/pub/job/{job_name}/{job_version}` - in this case they share single default Prometheus registry, so /metrics under this path will return also Job metrics. No, because for the root /metrics, they don't return the Job metrics. To fix this I tried to create custom singleton registry and save it in JNDI (sort of in-memory db) where it can be accessed by all servlets, however this was failing in some strange way, I believe because I also tried to use it in Valve (kind of Tomcat global filter for endpoints). Perhaps it's impossible (!) as the error message that I got `name [comp/env] is not bound in this context` I saw also in few questions on stack-overflow etc, with no solutions. Maybe without Valve it would work, but I think the way forward out of this is to implement https://github.com/TheRacetrack/plugin-tomcat-job-type/issues/10 , then attaching servlets to endpoints will be much easier/sensible and possibly also sharing single registry.

Reply:
This sounds like one of the Java nightmares. In this case, I guess the best way would be to have a one servlet inside a job type , handling all the paths and calling a class provided by a Job (maybe using reflection or something), but that's a story for another issue.

Reply:
But still that servlet wouldn't share prometheus registry with the root servlet (because of different classloader), so both /metrics endpoints would serve possibly different metrics. I really feel I've done everything here with current approach and came to wall, and possible solution is the mentioned rewrite.

Reply:
Java code becomes legacy and needs a rewrite few seconds after doing "git push" :smile: